scheduler/test/cook/test/kubernetes/controller.clj

```diff
@@ -38,39 +38,62 @@
+      (is (= :cook-expected-state/completed (do-process :cook-expected-state/running :pod/succeeded)))
+      (is (= :cook-expected-state/completed (do-process :cook-expected-state/running :pod/failed)))
       (is (= :cook-expected-state/running (do-process :cook-expected-state/running :pod/running)))
       (is (= :cook-expected-state/completed (do-process :cook-expected-state/running :pod/unknown)))
       (is (= :cook-expected-state/completed (do-process :cook-expected-state/running :pod/waiting)))
 
       (is (= :cook-expected-state/starting (do-process :cook-expected-state/starting :missing)))
-      (is (nil? (do-process :cook-expected-state/starting :pod/succeeded)))
-      (is (nil? (do-process :cook-expected-state/starting :pod/failed)))
+      (is (= :cook-expected-state/completed (do-process :cook-expected-state/starting :pod/succeeded)))
+      (is (= :cook-expected-state/completed (do-process :cook-expected-state/starting :pod/failed)))
       (is (= :cook-expected-state/running (do-process :cook-expected-state/starting :pod/running)))
       (is (= :cook-expected-state/completed (do-process :cook-expected-state/starting :pod/unknown)))
       (is (= :cook-expected-state/starting (do-process :cook-expected-state/starting :pod/waiting)))
 
       (is (nil? (do-process :missing :missing)))
       (is (nil? (do-process :missing :pod/succeeded)))
       (is (nil? (do-process :missing :pod/failed)))
-      (is (nil? (do-process :missing :pod/running)))
+      (is (= :missing (do-process :missing :pod/running)))
       (is (nil? (do-process :missing :pod/unknown)))
-      (is (nil? (do-process :missing :pod/waiting))))))
+      (is (= :missing (do-process :missing :pod/waiting))))))
+
+(deftest test-completion-protocol
+  (let [name "TestPodName"
+        do-process (fn [cook-expected-state k8s-actual-state]
+                     (let [cook-expected-state-map
+                           (atom {name {:cook-expected-state cook-expected-state}})]
+                       (controller/process
+                         {:api-client nil
+                          :cook-expected-state-map cook-expected-state-map
+                          :k8s-actual-state-map (atom {name {:synthesized-state {:state k8s-actual-state} :pod nil}})}
+                         name)
+                       (:cook-expected-state (get @cook-expected-state-map name {}))))
+        count-delete-pod (atom 0)]
+    (with-redefs [controller/delete-pod  (fn [_ cook-expected-state-dict _] (swap! count-delete-pod inc) cook-expected-state-dict)
+                  controller/log-weird-state (fn [_ _ _ _] :illegal_return_value_should_be_unused)
+                  controller/handle-pod-completed (fn [_ _] {:cook-expected-state :cook-expected-state/completed})
+                  controller/write-status-to-datomic (fn [_] :illegal_return_value_should_be_unused)]
+
+      (is (= :cook-expected-state/completed (do-process :cook-expected-state/killed :pod/succeeded)))
```

Comment:
I thought our "starting" state pair for this bug fix is (completed, succeeded), not (killed, succeeded)?

Reply:
Yeah. Good catch.

Reply:
How was this test passing when the starting state pair was wrong? Wouldn't we want the test to fail if we accidentally start it with the wrong state pair?

Reply:
When you're in killed,succeeded you transition to (completed,succeeded), and do the deletion protocol, which is what we measure. That intermediate state was invisible because of the recur.

Reply:
OK